librz/bin/p/bin_elf.inc

```diff
@@ -13,13 +13,351 @@
+};
+
+static Sdb *get_sdb(RzBinFile *bf) {
+	rz_return_val_if_fail(bf && bf->o && bf->o->bin_obj, NULL);
+
+	ELFOBJ *bin = rz_bin_file_get_elf(bf);
+	return bin->kv;
+}
+
+#ifndef RZ_BIN_CGC
+static int check_buffer_aux(RzBuffer *buf) {
+	rz_return_val_if_fail(buf, ELFCLASSNONE);
+
+	ut8 tmp[SELFMAG + 1] = { 0 };
+
+	if (rz_buf_read_at(buf, 0, tmp, SELFMAG + 1) < 0) {
+		return ELFCLASSNONE;
+	}
+
+	if (memcmp(tmp, ELFMAG, SELFMAG)) {
+		return ELFCLASSNONE;
+	}
+
+	return tmp[SELFMAG];
+}
+#endif
+
+static bool load_buffer(RZ_UNUSED RzBinFile *bf, RzBinObject *obj, RzBuffer *buf, RZ_UNUSED Sdb *sdb) {
+	rz_return_val_if_fail(obj, NULL);
+
+	ELFOBJ *bin = Elf_(rz_bin_elf_new_buf)(buf, &obj->opts);
+	if (!bin) {
+		return false;
+	}
+
+	obj->bin_obj = bin;
+	return true;
+}
+
+static ut64 baddr(RzBinFile *bf) {
+	rz_return_val_if_fail(bf && bf->o, UT64_MAX);
+
+	ELFOBJ *bin = rz_bin_file_get_elf(bf);
+	return Elf_(rz_bin_elf_get_baddr)(bin);
+}
+
+static ut64 boffset(RzBinFile *bf) {
+	rz_return_val_if_fail(bf && bf->o, UT64_MAX);
+
+	ELFOBJ *bin = rz_bin_file_get_elf(bf);
+	return Elf_(rz_bin_elf_get_boffset)(bin);
+}
+
+static ut64 binsym_aux(ELFOBJ *bin, RzBinSpecialSymbol sym) {
+	for (size_t i = 0; i < RZ_ARRAY_SIZE(special_symbol_translation_table); i++) {
+		if (sym == special_symbol_translation_table[i].symbol) {
+			return special_symbol_translation_table[i].get_addr(bin);
+		}
+	}
+
+	return UT64_MAX;
+}
+
+static RzBinAddr *rz_bin_addr_new_aux(ELFOBJ *bin, ut64 paddr, ut64 vaddr) {
+	RzBinAddr *result = RZ_NEW0(RzBinAddr);
+	if (!result) {
+		return NULL;
+	}
+
+	result->paddr = paddr;
+	result->vaddr = vaddr;
+	result->bits = bin->bits;
+
+	if (Elf_(rz_bin_elf_is_arm_binary_supporting_thumb)(bin)) {
+		rz_bin_elf_fix_arm_thumb_object_dispatch(result);
+	}
+
+	return result;
+}
+
+static RzBinAddr *rz_bin_addr_new_from_paddr(ELFOBJ *bin, ut64 paddr) {
+	ut64 vaddr = Elf_(rz_bin_elf_p2v_new)(bin, paddr);
+	if (vaddr == UT64_MAX) {
+		vaddr = paddr;
+	}
+
+	return rz_bin_addr_new_aux(bin, paddr, vaddr);
+}
+
+static RzBinAddr *rz_bin_addr_new_from_vaddr(ELFOBJ *bin, ut64 vaddr) {
+	ut64 paddr = Elf_(rz_bin_elf_v2p_new)(bin, vaddr);
+
+	return rz_bin_addr_new_aux(bin, paddr, vaddr);
+}
+
+static RzBinAddr *binsym(RzBinFile *bf, RzBinSpecialSymbol sym) {
+	rz_return_val_if_fail(bf && bf->o && bf->o, NULL);
+
+	ELFOBJ *bin = rz_bin_file_get_elf(bf);
+
+	ut64 paddr = binsym_aux(bin, sym);
+	if (paddr == UT64_MAX) {
+		return NULL;
+	}
+
+	return rz_bin_addr_new_from_paddr(bin, paddr);
+}
+
+static void add_constructor(ELFOBJ *bin, ut64 addr, ut64 size, int type, RzList *result) {
+	ut64 constructor_offset = Elf_(rz_bin_elf_v2p_new)(bin, addr);
+	if (constructor_offset == UT64_MAX) {
+		return;
+	}
+
+	ut64 pos = 0;
+
+	while (pos < size) {
+		ut64 offset = constructor_offset + pos;
+
+		Elf_(Addr) vaddr;
+		// TODO There is no way to see the diff between a value of UT64_MAX and an ERROR
```

Comment:
isn't this possible now?

Reply:
It should be possible now. I just need to remember why UT64_MAX is important.

Reply:
I have no idea why this comment was added… I removed the comment.

Reply:
This part should be moved to elf plugin later.